Termination overhaul

Piccions opened the pull request:
In this PR we improve the termination module:
- consensus of the termination is now done properly by relying on the GVT algorithm. Apart from implementation bugs, no deadlock should be possible.
-  the stable predicate termination detection doesn't rely anymore on hacky thread local variables, rather it is computed and committed on a per-LP basis
- we propose a new API for termination (I expect useful discussions here), which ideally would replace the handler based termination detection

In addition we introduce many smaller changes:
- the buddy system metadata memory footprint has been halved by relying on nibbles instead of bytes for metadata storage
- the autonomic checkpoint module now uses smaller width numbers (the LP struct now fits in two cache lines)
- many improvements in documentation and identifiers, which are being slowly migrated to less cryptic names
- decouple message processing from message extraction

This PR is a blocker for the rebinding and migration features, which cannot be easily implemented with the old termination module (which relies, as mentioned, on hacky thread local variable that encode LP metadata).

Piccions committed:
Reduce footprint of buddy metadatas, renamed index variables

Variables called index have been renamed to idx to
avoid conflicting with the homonym POSIX function

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions committed:
Reduce footprint of autonomic checkpoint interval module

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions committed:
Small gvt code improvement, remove mpi_remote_msg_drain()

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions committed:
Decouple process_msg() from extraction, improve termination detection

Also, renamed parallel.h in timewarp.h

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions committed:
Overhaul of the termination module and many smaller changes

The termination module has been completely overhauled. Now
the approach is the correct one, in which we make use of
the GVT algorithm to ensure consistent and logically
synchronized termination of the simulation even in a
distributed simulation.

Also, I propose a new termination API (which will need
further iterations before stabilization) which, in my
mind, should substitute the handler-based termination
detection.

In addition:
- improved debug code related to event scheduling
- removed m_seq from struct lp_msg (merged with flags)
- improved safety of tagged pointers in process_ctx
- improved anti message matching in distributed runs
- some documentation efforts

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions commented:
The CI failures on MacOS are due to the fact that, apparently, ```nanosleep()``` is *extremely* unreliable which means that a sleep of few milliseconds can take several seconds (https://stackoverflow.com/q/22784886). This has been fixed.

The CI failures on Windows are a mystery for me, for this I may require your help @alessandropellegrini

Piccions commented:
CI problems fixed, please give feedback on the changes!
In a recent discussion, @alessandropellegrini raised the concern about termination detection for unstable LP-local predicates.
I have in mind a solution that could implement the feature but:
- preliminary experiments showed that memory consumption would increase by at least 15%, as we would have to keep two GVT phases worth of snapshots and messages to ensure proper termination detection
- we don't have any real-world model that requires such feature, hence it would be hard to test the feature properly against common use cases, and its utility would be at best unclear

This more advanced feature, for the moment, could only be justified as a research toy, as such I believe it is outside of the scope of this PR.
Note that the CCGS approach used in ROOT-Sim 2 doesn't properly solve the problem for unstable predicates because it will miss transients between 2 GVT computations where all the LPs could potentially agree to complete.

alessandropellegrini commented:
Thanks for tracking our discussion here!

My concerns are actually twofold: on the one hand, I would like to avoid terminating the simulation _after_ the precise time at which all the LPs agreed to terminate, also in a stable termination predicate. This is what could happen when the termination is computed based on the GVT---but I admit that I am still behind with the code review, so I don't know if this is the case.

The second point is with the fact that an LP may change its mind based on the current state. So we should allow an LP to "unterminate" the simulation.

These two things together bring the problems that wasn't completely solved with CCGS, so what happens if you have an unstable predicate that's evaluated periodically. This last point has the highest cost, as you mentioned.

I have a mobility model somewhere that could benefit from this feature, and will share it with you as soon as I manage to find and polish it.

Clearly, reconstructing an aligned state brings overhead, and it shouldn't be enabled by default by any means!
A configuration option should make the job though: if a modeller needs this stuff, then they'll be willing to pay an additional cost.

alessandropellegrini commented on src/ROOT-Sim.h:
```diff
@@ -155,6 +156,28 @@ extern void rs_free(void *ptr);
+ * Its behavior, conditions for termination, or associated mechanisms (such as the committed handler) are being
+ * internally evaluated and may change in the near future. This warning will be removed when the feature will be stable.
+ */
+extern void rs_termination_request(void);
```
Two points on this API:

1. We typically use camel case for exported API. The exception is memory allocation API, because they were traditionally plugged in by the compiler/other stuff. This is no longer the case, and there is some inconsistency, but I would go with the camel case here.

2. Regarding the discussion on the retractability of the termination: we could use a parameter.

So this could be something like `HaltSimulation(bool)`?

But just to consider another option: do we really need a new API?
We may add a return value to `ProcessEvent`: `false`: don't halt; `true`: halt.

Clearly, this is changing the current API, and is possibly less explicative. So I have mixed feelings, and would like to hear from you.

Piccions replied:
Because of purely personal taste I prefer, API names in snake case with the ```rs_``` prefix. Anyway, maybe the problem is not relevant, for the moment; in light of our agreed upon changes to this PR, I think that we should have a better discussion on how to handle termination information from the user. So I plan on not adding any new API in this PR. This function will be removed in the next revision of the termination module!

alessandropellegrini commented on src/arch/timer.h:
```diff
@@ -66,7 +66,7 @@ static inline timer_uint timer_new(void)
 static inline timer_uint timer_value(timer_uint start)
 {
-	if(unlikely(timer_perf_freq == 0)) {
+	if(__builtin_expect(timer_perf_freq == 0, 0)) {
```
Why this change?

alessandropellegrini commented on src/distributed/control_msg.h:
```diff
@@ -10,17 +10,18 @@
-	MSG_CTRL_TERMINATION
+	MSG_CTRL_LP_END_TERMINATION,
+	/// Used in broadcast to signal that local LPs can terminate
+	MSG_CTRL_PHASE_ORANGE_TERMINATION,
```
I like the "Mattern style" colors of messages, but don't they make the code base less readable?

Piccions replied:
These will be removed in the new termination module!

alessandropellegrini commented on src/distributed/control_msg.h:
```diff
@@ -10,17 +10,18 @@
+	/// Used in broadcast to signal that local LPs can terminate
+	MSG_CTRL_PHASE_ORANGE_TERMINATION,
+	/// Used in broadcast to signal that local LPs can terminate
+	MSG_CTRL_PHASE_PURPLE_TERMINATION
```
From the comments, they seem to do the same stuff

Piccions replied:
These will be removed in the new termination module!

alessandropellegrini commented on src/distributed/mpi.c:
```diff
@@ -279,11 +230,10 @@ bool mpi_reduce_sum_scatter_done(void)
  *
  * Each node supplies a single simtime_t value. The minimum of all these values is computed and stored in @a node_min_p
- * itself. It is expected that only a single thread calls this function at a time. Each node has to call this function
+ * itself. It is expected that only a single thread calls this function at a time. Each rank has to call this function
```
Why don't we continue avoiding MPI terminology? IIRC, that's why you introduced the concept of nid?

Piccions replied:
I agree, using MPI terminology only makes sense when discussing MPI specific code.

alessandropellegrini commented on src/lp/process.h:
```diff
@@ -13,10 +13,28 @@
+/**
+ * The type of a Past Event Set entry
+ * Note that the exact same message buffer can appear as part of the PES of two different LPs. */
+enum pes_entry_type {
```
I am not sure if the comments to these values are correct?

alessandropellegrini commented on src/mm/buddy/buddy.c:
```diff
@@ -45,38 +45,28 @@ void buddy_init(struct buddy_state *self)
 	/* search recursively for the child */
-	uint_fast8_t node_size = B_TOTAL_EXP;
-	uint_fast32_t idx = 0;
+	uint_fast8_t node_size = B_TOTAL_EXP; // i = z - 1  z = i + 1
```
Is the comment a leftover?

Piccions commented:
Having a model that would benefit from the feature would be really useful, actually it is really important for me to test the feature.

Note that CCGS was unable to stop the simulation exactly at the first logical time that satisfies the termination condition (both for stable and unstable predicates) because, as you mentioned, it would run termination checks only at GVTs.
I have a solution for this but it costs additional memory. It wouldn't require to _periodically_ align the states but, as mentioned, it would require to keep more GVT periods worth of snapshots and messages. As you guessed, state alignment is currently not implemented.

So, if we agree, there should be a condition-based termination setting (an enum), with three values:
- disabled
- stable predicate
- unstable predicate

If non disabled, condition-based termination should align the LPs states to the first valid termination timestamp.
I would also suggest to deprecate the ```CanEnd()``` interface, and rather provide explicit functions that can be called by the LP to signal its termination state while evaluating the event handler.
Do you agree @alessandropellegrini ?

I will convert this PR to the draft state and continue working on it.

alessandropellegrini commented on src/lp/msg.h:
```diff
@@ -63,14 +63,20 @@ struct lp_msg {
 	union {
 		/// The flags to handle local anti messages
-		_Atomic uint32_t flags;
+		_Atomic uint64_t flags;
```
Now that the flags are in charge of many different things, I am wondering if using a bit field would increase the readability of the code base. We have plenty of masks here and there, many times without even a define. What do you think?

alessandropellegrini commented on src/lp/process.h:
```diff
@@ -13,10 +13,28 @@
+/// An entry in the Past Event Set (see struct process_ctx)
+struct pes_entry {
+	/// A tagged pointer used to distinguish between different message kinds (see enum pes_entry)
+	uintptr_t raw;
```
Shall we consider using bit fields also here?

alessandropellegrini commented on src/mm/buddy/buddy.c:
```diff
@@ -93,40 +83,28 @@ void *buddy_malloc(struct buddy_state *self, const uint_fast8_t req_blks_exp)
-		bitmap_set(self->dirty, i >> B_BLOCK_EXP);
-#endif
+	for(idx >>= 1; idx; idx >>= 1) {
+		uint_fast8_t left_long = self->longest[idx] & 0x0f;
+		uint_fast8_t right_long = self->longest[idx] >> 4;
+
+		uint_fast8_t c;
+		if(left_long == node_size && right_long == node_size)
+			c = node_size + 1;
+		else
+			c = max(left_long, right_long);
+
+		uint_fast8_t d = self->longest[idx >> 1U];
+		self->longest[idx >> 1U] = idx & 1U ? (d & 0xf0) | c : (d & 0x0f) | (c << 4U);
```
Maybe we can use a define for `0x0f` and `0xf0`? I am thinking about poor prospective students reading this code :smile:

Piccions replied:
I will think of some ways to make this code actually readable. I am having a look right now; honestly, after a while not reading this, it's brutal :sob:

alessandropellegrini commented on src/mm/buddy/checkpoint.c:
```diff
@@ -23,26 +23,28 @@
 				on_visit(__o, __len);                                                                  \
-			} else if(__lon != __l) {                                                                      \
-				__i = buddy_left_child(__i) + __vis;                                                   \
+			} else if(__lon + B_BLOCK_EXP - 1 != __l) {                                                    \
+				__i = (__i << 1U) + __vis;                                                             \
 				__vis = false;                                                                         \
 				__l--;                                                                                 \
 				continue;                                                                              \
 			}                                                                                              \
 			do {                                                                                           \
-				__vis = !(__i & 1U);                                                                   \
-				__i = buddy_parent(__i);                                                               \
+				__vis = __i & 1U;                                                                      \
+				__i >>= 1U;                                                                            \
```
Leftover trailing spaces?

alessandropellegrini commented on src/mm/checkpoint/autonomic.c:
```diff
@@ -24,9 +24,9 @@
-		o *(((f) - 1.0) / (f)) + s * (1.0 / (f));                                                              \
+		float s = (sample);                                                                                    \
+		float o = (old_v);                                                                                     \
+		(((f) - 1.0f) / (f)) * o + (1.0f / (f)) * s;                                                           \
```
:+1:

alessandropellegrini commented:
I have completed my review and left some discussion comments in there.

> So, if we agree, there should be a condition-based termination setting (an enum), with three values:
> 
> * disabled
> * stable predicate
> * unstable predicate

I mostly agree. I am only doubtful about `disabled`: simulation should terminate somehow. Are you referring to the timestamp-based termination in that case?
If so, I am also unsure: I may want to terminate the simulation either based on a reached VT, or based on a predicate.

So, probably my concern is if the termination time is 0 and condition-based termination is disabled. We could raise either a warning or an error in that case---probably a warning is safer.

> If non disabled, condition-based termination should align the LPs states to the first valid termination timestamp. I would also suggest to deprecate the `CanEnd()` interface, and rather provide explicit functions that can be called by the LP to signal its termination state while evaluating the event handler. Do you agree @alessandropellegrini ?

Yes, sure, let's deprecate for now `CanEnd()`. Let's not remove it yet though, as this would require bumping major version number, which is probably something we should defer---there are some APIs that are getting deprecated now, and it would be ok to remove them all at once.
> 
> I will convert this PR to the draft state and continue working on it.

Thanks a lot!!!

Piccions committed:
Dirty commit: to cleanup and rebase!

Criticism and suggestions are always welcome!

Signed-off-by: Andrea Piccione <piccione@diag.uniroma1.it>

Piccions commented:
> I mostly agree. I am only doubtful about `disabled`: simulation should terminate somehow. Are you referring to the timestamp-based termination in that case? If so, I am also unsure: I may want to terminate the simulation either based on a reached VT, or based on a predicate.
> 
> So, probably my concern is if the termination time is 0 and condition-based termination is disabled. We could raise either a warning or an error in that case---probably a warning is safer.

In my mind, condition-based termination and virtual time based termination are orthogonal (the user should be able to enable/disable them independently). A warning should be more than sufficient in case both of them are disabled; a simulation could still terminate because of no messages being scheduled!

Please, when you can, let me have a model that would benefit from the feature!

Piccions commented:
I have a (allegedly) working implementation of the per-LP unstable predicate termination detection. When not used, the overhead is minimal, and even when used it doesn't look too bad, but I can't know for sure until I have a model that makes a meaningful use of the feature. Do you think you can provide me with the mobility model you mentioned some time ago? @alessandropellegrini

alessandropellegrini committed:
Add mobility model

This commit introduces a mobility model with unstable termination condition, as a test.
The test may currently run indefinitely :-)

Signed-off-by: Alessandro Pellegrini <a.pellegrini@ing.uniroma2.it>
